internal/hcsoci/assigned_devices.go

```diff
@@ -0,0 +1,157 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+
+	winio "github.com/Microsoft/go-winio"
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+func specHasAssignedDevices(coi *createOptionsInternal) bool {
+	if (coi.Spec.Windows != nil) && (coi.Spec.Windows.Devices != nil) &&
+		(len(coi.Spec.Windows.Devices) > 0) {
+		return true
+	}
+	return false
+}
+
+// handleAssignedDevicesWindows does all of the work to setup the hosting UVM and retrieve
+// device information for adding assigned devices on a WCOW container definition.
+//
+// First, devices are assigned into the hosting UVM. Drivers are then added into the UVM and
+// installed on the matching devices. This ordering allows us to guarantee that driver
+// installation on a device in the UVM is completed before we attempt to create a container.
+//
+// Then we find the location paths of the target devices in the UVM and return the results
+// as WindowsDevices.
+func handleAssignedDevicesWindows(ctx context.Context, coi *createOptionsInternal, r *Resources) ([]specs.WindowsDevice, error) {
+	vpciVMBusInstanceIDs := []string{}
+	for _, d := range coi.Spec.Windows.Devices {
+		if d.IDType == uvm.VPCIDeviceIDType || d.IDType == uvm.VPCIDeviceIDTypeLegacy {
+			vpci, err := coi.HostingSystem.AssignDevice(ctx, d.ID)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to assign device %s of type %s to pod %s", d.ID, d.IDType, coi.HostingSystem.ID())
+			}
+			r.resources = append(r.resources, vpci)
+			vmBusInstanceID := coi.HostingSystem.GetAssignedDeviceParentID(vpci.VMBusGUID)
+			log.G(ctx).WithField("vmbus id", vmBusInstanceID).Info("vmbus instance ID")
+
+			vpciVMBusInstanceIDs = append(vpciVMBusInstanceIDs, vmBusInstanceID)
+		}
+	}
+	if len(vpciVMBusInstanceIDs) == 0 {
+		return nil, fmt.Errorf("no assignable devices on the spec %v", coi.Spec.Windows.Devices)
```

Comment:
I think we will produce better diagnostics if we instead error when we find an unsupported device in `coi.Spec.Windows.Devices`, rather than ignoring it and then erroring because our list of added devices is empty.

Reply:
Done